### PR TITLE
best metric log with wandb

### PR DESCRIPTION
**Pull request by okotaku**
## Motivation

Log best metric/loss with wandb.

## Use cases (Optional)

```
log_config = dict(
    interval=50,
    ignore_last=False,
    hooks=[
        dict(type='TextLoggerHook'),
        dict(type='WandbLoggerHook',
             best_log_dict={'loss': 'min'},
             init_kwargs=dict(
                 project='example',
                 name='example001'
             ))
    ])
```

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/runner/hooks/logger/wandb.py [hunk 1]
```diff
@@ -87,6 +93,17 @@ def before_run(self, runner):
     @master_only
     def log(self, runner):
         tags = self.get_loggable_tags(runner)
+
+        # log best metric
+        for k, v in list(tags.items()):
```
Comment (HAOCHENYE) [inline mmcv/runner/hooks/logger/wandb.py:98]: Single letter variable is hard to understand, could you rename k(tag_keys), v(tag_values) bk(metric_keys) and bv(best_metric_value)?

File: mmcv/runner/hooks/logger/wandb.py [hunk 2]
```diff
@@ -87,6 +93,17 @@ def before_run(self, runner):
...
+            for bk, bv in self.best_log_dict.items():
+                if bk in k:
+                    if (k not in self.best_metrics) or (
```
Comment (HAOCHENYE) [inline mmcv/runner/hooks/logger/wandb.py:101]: This `if()` is too long to understand, maybe you can write them separately.

File: mmcv/runner/hooks/logger/wandb.py [hunk 3]
```diff
@@ -87,6 +93,17 @@ def before_run(self, runner):
...
+                        (bv == 'max') and (v > self.best_metrics[k])) or (
+                            (bv == 'min') and (v < self.best_metrics[k])):
+                        self.best_metrics[k] = v
+                        tags[f'{k}_best'] = v
+
```
Comment (HAOCHENYE) [inline mmcv/runner/hooks/logger/wandb.py:106]: It would be nice if you could add unit tests.

**Review by HAOCHENYE — APPROVED** [review]
LGTM.

**Comment by okotaku** [comment]
I think so.
If you save the best loss/metric, you can compare them as a list on wandb.

スクリーンショット 2022-01-31 8 56 47

**Comment by zhouzaida** [comment]
> I think so. If you save the best loss/metric, you can compare them as a list on wandb. (Image left) Image right is last epoch loss/metric.
> 
> スクリーンショット 2022-01-31 9 02 39

Got it.

**Comment by zhouzaida** [comment]
Hi @okotaku , thanks for your contribution. It is a useful feature but not a very common feature, so we are not going to merge it into master branch.

**Pull request closed without merging**